src/data/finsum.lean

```diff
@@ -0,0 +1,657 @@
+      refine ⟨_, hx₁⟩,
+      by_cases hf : f x = 0,
+      { rw [function.mem_support, pi.add_apply, hf, zero_add] at hx₂,
+        apply finset.subset_union_right,
+        rw set.finite.mem_to_finset,
+        exact ⟨hx₁, hx₂⟩ },
+      { apply finset.subset_union_left,
+        rw set.finite.mem_to_finset,
+        exact ⟨hx₁, hf⟩ } },
+    { rintros x ⟨hxs, hx⟩,
+      rw [finset.mem_coe, finset.mem_filter] at hxs,
+      by_contra hfg, exact hx ⟨hxs.2, hfg⟩ } },
+  all_goals { rw [← finsum_in_eq_finset_sum'''', finsum_in_inter_support],
+              apply finsum_in_eq_finsum_in_of_subset;
+              { intro x, finish } },
+end .
+
+/-- Given the support of `f` and `g` are finite, the sum over `f + g` equals the sum over `f` plus
+  the sum over `g`. -/
+lemma finsum_add_distrib
+  (hf : (function.support f).finite) (hg : (function.support g).finite) :
+  ∑ᶠ i, (f + g) i = ∑ᶠ i, f i + ∑ᶠ i, g i :=
+begin
+  rw [finsum_eq_finsum_in_univ, finsum_eq_finsum_in_univ, finsum_eq_finsum_in_univ],
+  apply finsum_in_add_distrib',
+  { convert hf, simp },
+  { convert hg, simp }
+end
+
+/-- Given the function `f : α → β → M`, the finite sets `s : set α`, `t : set β`, the sum on `s`
+  over `λ x, finsum_in t (λ y, f x y)` equals the sum on `t` over `λ y, finsum_in s (λ x, f x y)`.
+  -/
+lemma finsum_in_comm {s : set α} {t : set β}
+  (f : α → β → M) (hs : s.finite) (ht : t.finite) :
+  ∑ᶠ i in s, ∑ᶠ j in t, f i j = ∑ᶠ j in t, ∑ᶠ i in s, f i j :=
+begin
+  rw [finsum_in_eq_finset_sum''' _ hs, finsum_in_eq_finset_sum''' _ ht],
+  conv_lhs { congr, skip, funext, rw finsum_in_eq_finset_sum''' _ ht },
+  conv_rhs { congr, skip, funext, rw finsum_in_eq_finset_sum''' _ hs },
+  exact finset.sum_comm,
+end
+
+/-- Given a monoid homomorphism `g : β → M`, and a function `f : α → β`, the sum on `s` over `g ∘ f`
+  equals `g` of the sum on `s` over `f`. -/
+lemma finsum_in_hom [add_comm_monoid β] (f : α → β) (g : β →+ M) (hs : s.finite) :
+  ∑ᶠ i in s, (g ∘ f) i = g (∑ᶠ j in s, f j) :=
+by rw [finsum_in_eq_finset_sum''' _ hs, finsum_in_eq_finset_sum''' _ hs,
+       set.finite.to_finset, finset.sum_hom']
+
+lemma finsum_in_hom'' [add_comm_monoid β] (f : α → β) (g : β →+ M) (hs : s.finite) :
+  ∑ᶠ i in s, g (f i) = g (∑ᶠ j in s, f j) :=
+finsum_in_hom f g hs
+
+/-- A more general version of `finsum_in_hom` that requires `s ∩ function.support f` and instead of
+  `s` to be finite. -/
+lemma finsum_in_hom' [add_comm_monoid β] {f : α → β} (g : β →+ M)
+  (h₀ : (s ∩ function.support f).finite) :
+  ∑ᶠ i in s, (g ∘ f) i = g (∑ᶠ j in s, f j) :=
+begin
+  rw [finsum_in_inter_support f, ← finsum_in_hom _ _ h₀],
+  symmetry,
+  refine finsum_in_eq_finsum_in_of_subset'' (set.inter_subset_left _ _)
+    (set.inter_subset_inter_right _ (λ x hgx, _)),
+  rw function.mem_support at hgx ⊢,
+  intro hfx, refine hgx _, simp only [hfx, function.comp],
+  exact add_monoid_hom.map_zero g
+end
+
+lemma finsum_in_hom''' [add_comm_monoid β] {f : α → β} (g : β →+ M)
```

Comment:
This version should replace `finsum_in_hom'`

Reply:
Done